src/engine/strat_engine/metadata.rs

```diff
@@ -4,262 +4,416 @@
+            reserved_size: Sectors(LittleEndian::read_u64(&buf[104..112]) as u64),
+            flags: 0,
+        })
+    }
+
+    /// Read pool uuid from buf at offset.
+    /// Return an error if the data can't be interpreted as a Uuid.
+    pub fn read_pool_uuid(buf: &[u8; BDA_STATIC_HDR_SIZE as usize]) -> EngineResult<Uuid> {
+        Uuid::parse_str(from_utf8(&buf[SECTOR_SIZE as usize + 32..SECTOR_SIZE as usize + 64])
+                .unwrap())
+            .map_err(|_| EngineError::Engine(ErrorEnum::Invalid, "invalid pool uuid".into()))
     }
 
-    /// Return the least recently timestamped MDA.
-    pub fn least_recent(&mut self) -> &mut MDA {
-        match self.mdaa.last_updated.cmp(&self.mdab.last_updated) {
-            Ordering::Less => &mut self.mdaa,
-            Ordering::Greater => &mut self.mdab,
-            _ => &mut self.mdaa,
+    /// Determine the ownership of a device based on data in buf.
+    pub fn determine_ownership(buf: &[u8; BDA_STATIC_HDR_SIZE as usize])
+                               -> EngineResult<DevOwnership> {
+        if &buf[SECTOR_SIZE as usize + 4..SECTOR_SIZE as usize + 20] == STRAT_MAGIC {
+            Ok(DevOwnership::Ours(try!(StaticHeader::read_pool_uuid(&buf))))
+        } else {
+            if buf.iter().any(|x| *x != 0) {
+                Ok(DevOwnership::Theirs)
+            } else {
+                Ok(DevOwnership::Unowned)
+            }
         }
     }
 }
 
 #[derive(Debug, Clone)]
-pub struct SigBlock {
-    pub total_size: Sectors,
-    pub pool_uuid: Uuid,
-    pub dev_uuid: Uuid,
-    pub mda: MDAGroup,
-    pub mda_sectors: Sectors,
-    pub reserved_sectors: Sectors,
+pub struct MDARegions {
+    // Spec defines 4 regions, but regions 2 & 3 are duplicates of 0 and 1 respectively
+    pub region_size: Sectors,
+    pub mdas: [MDAHeader; 2],
 }
 
-impl SigBlock {
-    pub fn new(pool_uuid: &Uuid,
-               dev_uuid: &Uuid,
-               mda_size: Sectors,
-               total_size: Sectors)
-               -> SigBlock {
-        SigBlock {
-            mda: MDAGroup::new(mda_size),
-            pool_uuid: pool_uuid.clone(),
-            dev_uuid: dev_uuid.clone(),
-            total_size: total_size,
-            mda_sectors: mda_size,
-            reserved_sectors: MDA_RESERVED_SIZE,
+impl MDARegions {
+    pub fn initialize(header: &StaticHeader, f: &mut File) -> EngineResult<MDARegions> {
+        let region_size = Sectors(*header.mda_size / NUM_MDA_REGIONS);
+        let per_region_size: u64 = *region_size * SECTOR_SIZE;
+        let hdr_buf = [0u8; MDA_REGION_HDR_SIZE];
+
+        for region in 0..NUM_MDA_REGIONS {
+            try!(f.seek(SeekFrom::Start(BDA_STATIC_HDR_SIZE + (region as u64 * per_region_size))));
+            try!(f.write_all(&hdr_buf));
+        }
+
+        try!(f.flush());
+
+        Ok(MDARegions {
+            region_size: region_size,
+            mdas: [MDAHeader::new(per_region_size), MDAHeader::new(per_region_size)],
```

Comment:
Seems wierd to me that each header has a ```per_region_size``` attribute, since I think they are the same for all headers. But it does look more convenient.